keystorage/src/main/java/tech/pegasys/web3signer/keystorage/awskms/AwsKeyManagerService.java

```diff
@@ -0,0 +1,172 @@
+  private final AwsAuthenticationMode authMode;
+  private final AwsCredentials awsCredentials;
+  private final String region;
+
+  private KmsClient kmsClient;
+
+  private Optional<URI> endpointOverride;
+
+  public AwsKeyManagerService(
+      final AwsAuthenticationMode authMode,
+      final AwsCredentials awsCredentials,
+      final String region,
+      final Optional<URI> endpointOverride) {
+    this.authMode = authMode;
+    this.awsCredentials = awsCredentials;
+    this.region = region;
+    this.endpointOverride = endpointOverride;
+
+    initKmsClient();
+  }
+
+  void initKmsClient() {
+    if (kmsClient != null) {
+      return;
+    }
+
+    final KmsClientBuilder kmsClientBuilder = KmsClient.builder();
+    endpointOverride.ifPresent(kmsClientBuilder::endpointOverride);
+    kmsClient =
+        kmsClientBuilder
+            .credentialsProvider(getAwsCredentialsProvider())
+            .region(Region.of(region))
+            .build();
+  }
+
+  private AwsCredentialsProvider getAwsCredentialsProvider() {
+    final AwsCredentialsProvider awsCredentialsProvider;
+    if (authMode == ENVIRONMENT) {
+      awsCredentialsProvider = DefaultCredentialsProvider.create();
+    } else {
+      final software.amazon.awssdk.auth.credentials.AwsCredentials awsSdkCredentials;
+      if (awsCredentials.getSessionToken().isPresent()) {
+        awsSdkCredentials =
+            AwsSessionCredentials.create(
+                awsCredentials.getAccessKeyId(),
+                awsCredentials.getSecretAccessKey(),
+                awsCredentials.getSessionToken().get());
+      } else {
+        awsSdkCredentials =
+            AwsBasicCredentials.create(
+                awsCredentials.getAccessKeyId(), awsCredentials.getSecretAccessKey());
+      }
+
+      awsCredentialsProvider = StaticCredentialsProvider.create(awsSdkCredentials);
+    }
+    return awsCredentialsProvider;
+  }
+
+  public ECPublicKey getECPublicKey(final String kmsKeyId) {
+    checkArgument(kmsClient != null, "KmsClient is not initialized");
+
+    // Question ... do we need to set grantTokens?
+    final GetPublicKeyRequest getPublicKeyRequest =
+        GetPublicKeyRequest.builder().keyId(kmsKeyId).build();
+    final GetPublicKeyResponse publicKeyResponse = kmsClient.getPublicKey(getPublicKeyRequest);
+    KeySpec keySpec = publicKeyResponse.keySpec();
+    if (keySpec != KeySpec.ECC_SECG_P256_K1) {
+      throw new RuntimeException("Unsupported key spec from AWS KMS: " + keySpec.toString());
+    }
+
+    final X509EncodedKeySpec encodedKeySpec =
+        new X509EncodedKeySpec(publicKeyResponse.publicKey().asByteArray());
+    try {
+      KeyFactory keyFactory = KeyFactory.getInstance("EC");
+      return (ECPublicKey) keyFactory.generatePublic(encodedKeySpec);
+    } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public byte[] sign(final String kmsKeyId, final byte[] data) {
```

Comment:
For Azure and file-based, the sign operation is done at the Signer level of abstraction (in the signing package) - is there a reason this one is pushed down to key vault level of abstraction?

Reply:
I think that we should not expose the actual Azure or AWS library outside of keystorage module. There is additional logic around `sign`, such as converting to ethereum specific signature etc., that is correctly performed in signing module. `AzureKeyVaultSigner` and `Factory` should not directly use `CryptoClient` directly as well.